File: src/Statistics/java/src/test/java/SmartWordToyTest.java
import static org.junit.Assert.assertEquals;
import org.junit.Test;

class Params {
  String start;
  String finish;
  String[] forbid;
  int result;
  Params(String start, String finish, String[] forbid, int result){
      this.start=start;
      this.finish=finish;
      this.forbid=forbid;
      this.result=result;
  }
}

public class SmartWordToyTest {

  @Test
  public void one() {
      Params[] list = new Params[] {
        new Params(
              "aaaa",
              "zzzz",
              new String[]{"a a a z", "a a z a", "a z a a", "z a a a", "a z z z", "z a z z", "z z a z", "z z z a"},
              8
          ),
          new Params(
              "aaaa",
              "bbbb",
              new String[]{},
              4
          ),
          new Params(
              "aaaa",
              "mmnn",
              new String[]{},
              50
          ),
          new Params(
              "aaaa",
              "zzzz",
              new String[]{"bz a a a", "a bz a a", "a a bz a", "a a a bz"},
              -1
          ),
          new Params(
              "aaaa",
              "zzzz",
              new String[]{
                  "cdefghijklmnopqrstuvwxyz a a a",
                  "a cdefghijklmnopqrstuvwxyz a a",
                  "a a cdefghijklmnopqrstuvwxyz a",
                  "a a a cdefghijklmnopqrstuvwxyz"
              },
              6
          ),
          new Params(
              "aaaa",
              "bbbb",
              new String[]{"b b b b"},
              -1
          ),
          new Params(
              "zzzz",
              "aaaa",
              new String[]{
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                  "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk"
              },
              -1
          )
      };

    SmartWordToy toy = new SmartWordToy();

    for (int i = 0; i < list.length; i++) {
      int actual = toy.minPresses(list[0].start, list[0].finish, list[0].forbid);
      assertEquals(list[0].result, actual);
    }
  }
}

File: src/Statistics/SmartWordToyTest.php
<?php
/*
 * This file is part of the phporithms.
 *
 * (c) Alexander Golovnya <snsanich@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 * Date: 3/28/17
 * Time: 10:03 PM
 */

namespace Phporithms\Statistics;

class SmartWordToyTest extends \PHPUnit\Framework\TestCase
{
    public function provider()
    {
        return [
            [
                "aaaa",
                "zzzz",
                ["a a a z", "a a z a", "a z a a", "z a a a", "a z z z", "z a z z", "z z a z", "z z z a"],
                8
            ],
            [
                "aaaa",
                "bbbb",
                [],
                4
            ],
            [
                "aaaa",
                "mmnn",
                [],
                50
            ],
            [
                "aaaa",
                "zzzz",
                ["bz a a a", "a bz a a", "a a bz a", "a a a bz"],
                -1
            ],
            [
                "aaaa",
                "zzzz",
                [
                    "cdefghijklmnopqrstuvwxyz a a a",
                    "a cdefghijklmnopqrstuvwxyz a a",
                    "a a cdefghijklmnopqrstuvwxyz a",
                    "a a a cdefghijklmnopqrstuvwxyz"
                ],
                6
            ],
            [
                "aaaa",
                "bbbb",
                ["b b b b"],
                -1
            ],
            [
                "zzzz",
                "aaaa",
                [
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk",
                    "abcdefghijkl abcdefghijkl abcdefghijkl abcdefghijk"
                ],
                -1
            ]
        ];
    }

    /**
     * @dataProvider provider
     * @param string $start
     * @param string $finish
     * @param array $forbid
     * @param int $expected
     */
    public function testOne(string $start, string $finish, array $forbid, int $expected)
    {
        $toy = new SmartWordToyWithIdentifier;
        $this->assertEquals($expected, $toy->minPresses($start, $finish, $forbid));
    }
}
